Add variant generic interfaces introduced in 4.5 (#11753)
dotnet/docs: docs/csharp/programming-guide/concepts/covariance-contravariance/variance-in-generic-interfaces.md

> 

Reviewer: line 3 contains ms.date, the date this topic was last updated. Please change it to today's date.

Author: Yep, will do.

# Variance in Generic Interfaces (C#)

.NET Framework 4 introduced variance support for several existing generic interfaces. Variance support enables implicit conversion of classes that implement these interfaces. 

> The following interfaces are now variant:

Reviewer: ```suggestion
Staring with .NET Framework 4, the following interfaces are variant:
```

- <xref:System.IComparable%601> (T is contravariant)

> The following variant interfaces were added in .NET Framework 4.5:

Reviewer: ```suggestion
Starting with .NET Framework 4.5, the following interfaces are variant:
```

- <xref:System.Collections.Generic.IReadOnlyList%601> (T is contravariant)

- <xref:System.Collections.Generic.IReadOnlyCollection%601> (T is contravariant)

Covariance permits a method to have a more derived return type than that defined by the generic type parameter of the interface. To illustrate the covariance feature, consider these generic interfaces: `IEnumerable<Object>` and `IEnumerable<String>`. The `IEnumerable<String>` interface does not inherit the `IEnumerable<Object>` interface. However, the `String` type does inherit the `Object` type, and in some cases you may want to assign objects of these interfaces to each other. This is shown in the following code example.

> ```csharp

Reviewer: unrelated to your PR: on line 40, it's VB and not C# that has an `Option Strict On` setting. Could you change this to read "...a compilation error in C# and, if `Option Strict` is on, in Visual Basic."

Author: ah, good catch, will do.
